harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        free_mem_bytes = psutil.virtual_memory().available
+        return {
+            "free_memory": free_mem_bytes / 1e9,
+        }
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        cpu_util = psutil.cpu_percent()
+        return {
+            "cpu_util_simple": cpu_util,
+        }
+
+
+class _GPU(_MetricGroupCollector):
+    group = "gpu"
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._pynvml_device_handles: typing.Dict[str, typing.Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning("pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        """
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+        """
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: typing.Dict[str, typing.Any],
+        timestamp: datetime.datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        max_collection_s: int = 60 * 30,
```

Comment:
This is not configurable by the end user. Should it be?

Reply:
Does the default claim that if a Trial is run for more than half an hour, something is up and we don't care about metrics anymore?

Reply:
Found:

> Extend/remove current 5 minute limit on profiling
This exists to speed up load times on Web UI. The profiler was originally intended to capture training metrics at high granularity for a small amount of time.
Investigate appropriate granularity / time cap to preserve load times on Web UI.

Reply:
not for now, but wanted to make it option from an interface standpoint, and maybe for users at a late rpoint in time